server/src/game-implementations/games/maimai.ts

```diff
@@ -0,0 +1,233 @@
+import {
+	GoalFmtPercent,
+	GoalOutOfFmtPercent,
+	GradeGoalFormatter,
+} from "./_common";
+import db from "external/mongo/db";
+import { CreatePBMergeFor } from "game-implementations/utils/pb-merge";
+import { SessionAvgBest10For } from "game-implementations/utils/session-calc";
+import { MaimaiRate } from "rg-stats";
+import { GetGrade, MAIMAI_GBOUNDARIES } from "tachi-common";
+import { IsNullish } from "utils/misc";
+import type { GPTServerImplementation } from "game-implementations/types";
+import type {
+	ExtractedClasses,
+	Game,
+	integer,
+	Playtype,
+	UserGameStatsSnapshotDocument,
+} from "tachi-common";
+import { FindOneResult } from "monk";
+
+const DAN_BONUSES: Record<ExtractedClasses["maimai:Single"]["dan"], number> = {
+	"DAN_1": 0.2,
+	"DAN_2": 0.4,
+	"DAN_3": 0.55,
+	"DAN_4": 0.7,
+	"DAN_5": 0.85,
+	"DAN_6": 0.95,
+	"DAN_7": 1.05,
+	"DAN_8": 1.15,
+	"DAN_9": 1.25,
+	"DAN_10": 1.35,
+	"KAIDEN": 1.65,
+
+	"SHINDAN_1": 1.70,
+	"SHINDAN_2": 1.75,
+	"SHINDAN_3": 1.80,
+	"SHINDAN_4": 1.85,
+	"SHINDAN_5": 1.90,
+	"SHINDAN_6": 1.95,
+	"SHINDAN_7": 1.98,
+	"SHINDAN_8": 2.01,
+	"SHINDAN_9": 2.04,
+	"SHINDAN_10": 2.07,
+	"SHINKAIDEN": 2.10,
+};
+
+// maimai rating is composed of three components: BEST, RECENT, HISTORY
+// The BEST and RECENT components consist of N best **songs**, that is,
+// each song can only be in the BEST or RECENT component once.
+//
+// For example, if a player has a 99% on PANDORA PARADOXX Re:MASTER and
+// a 100% on PANDORA PARADOXX MASTER, only the MASTER score will be counted
+// since it is higher rated.
+//
+// There is also a dan bonus, up to 2.10 for Shinkaidan.
+//
+// This does not take into account the RECENT component.
+async function CalculateMaimaiRate(
+	game: Game,
+	playtype: Playtype,
+	userID: integer,
+) {
+	const user = await db["game-stats-snapshots"].findOne(
```

Comment:
This is the wrong place to fetch a users dan from. Use `game-stats` instead of `game-stats-snapshots`. The snapshots are purely for graphing purposes.

Reply:
marking as resolved since we moved entirely to naiveRating, no dans required.